### ISSUE TITLE
Client is always accepted for undeclared namespaces

### ISSUE THREAD

**Issue by llunn** [miguelgrinberg/python-socketio#822]
**Describe the bug**
This issue is manifesting itself through flask-socketio, but I believe the source of the issue belongs here. When a client connects using a namespace that is not declared, the namespace is simply added and the client connects with no ability to execute code against the connection in order to reject it.  

The end result is that it is possible to pollute the connection pool with phantom clients, and the server cannot detect when new clients connect. 

The source of issue I believe is here:  https://github.com/miguelgrinberg/python-socketio/blob/3bd13578c82e8a94d6b0328180606c2aefd496f1/src/socketio/base_manager.py#L119

**To Reproduce**

1. Stand up a bare bones server instance.
2. Define an on_connect handler for '/'
2. Define zero or more class based namespaces and register them
3. Connect a client using a namespace that the server is not aware of.
4. Client connection is blindly accepted, regardless of `always_connect` value.

**Expected behavior**
One of:

1. Connections to namespaces that do not exist are flat out rejected; or
1. Connections to namespaces that do not exist produce an exception; or
2. Connect event is delegated to to `/` namespace `connect`, with relevant details on the namespace attempted to connect to.

In all cases, it is also expected that the `/` namespace `connect` event fires, which is also not occurring.

**Logs**
```
(58685) wsgi starting up on http://0.0.0.0:5000
(58685) accepted ('127.0.0.1', 50507)
1_szfyPlEneUJwA3AAAA: Sending packet OPEN data {'sid': '1_szfyPlEneUJwA3AAAA', 'upgrades': [], 'pingTimeout': 20000, 'pingInterval': 25000}
kombu backend initialized.
1_szfyPlEneUJwA3AAAA: Received request to upgrade to websocket
1_szfyPlEneUJwA3AAAA: Upgrade to websocket successful
1_szfyPlEneUJwA3AAAA: Received packet MESSAGE data 0/celery,{"authToken":"bGVlQHRvcnVzb2Z0LmNvbTo2MUE1MUI1NTrMnz4SGR56ssXsYgxiiGHGa5ck6g"}
/celery
1_szfyPlEneUJwA3AAAA: Sending packet MESSAGE data 0/celery,{"sid":"hOsZtTCikRL7RTIfAAAB"}
(58685) accepted ('192.168.0.60', 50513)
DTnpUXYe8iuRM0UdAAAC: Sending packet OPEN data {'sid': 'DTnpUXYe8iuRM0UdAAAC', 'upgrades': [], 'pingTimeout': 20000, 'pingInterval': 25000}
DTnpUXYe8iuRM0UdAAAC: Received request to upgrade to websocket
DTnpUXYe8iuRM0UdAAAC: Upgrade to websocket successful
DTnpUXYe8iuRM0UdAAAC: Received packet MESSAGE data 0/remote,{"authToken":"bGVlQHRvcnVzb2Z0LmNvbTo2MUE1MUI1NTrMnz4SGR56ssXsYgxiiGHGa5ck6g","userCtx":{"name":"lee@torusoft.com","roles":["confide-analytics","emmit-admin","confide-admin"]}}
/remote
DTnpUXYe8iuRM0UdAAAC: Sending packet MESSAGE data 0/remote,{"sid":"DhB9khVI587ck8pjAAAD"}
(58685) accepted ('192.168.0.60', 50558)
```
**Additional context**
If this is not intended behaviour, a possible easy solution is to add an else to to the if else if block of [Server._trigger_event](https://github.com/miguelgrinberg/python-socketio/blob/main/src/socketio/server.py#L740) 

If more detail is required, please let me know!

**Comment by miguelgrinberg** [comment]
Yes, I believe your analysis is correct. The server should not accept a connection for a namespace it does not know about.

> In all cases, it is also expected that the / namespace connect event fires, which is also not occurring.

I don't understand what you mean here. The current version of the Socket.IO protocol does not treat the `/` namespace in a special way as it did in previous versions. Only the namespaces the client connects to are connected, `/` needs to be connected explicitly now.

**Comment by llunn** [comment]
@miguelgrinberg Thanks! 

> > In all cases, it is also expected that the / namespace connect event fires, which is also not occurring.
> 
> I don't understand what you mean here. The current version of the Socket.IO protocol does not treat the `/` namespace in a special way as it did in previous versions. Only the namespaces the client connects to are connected, `/` needs to be connected explicitly now.

Full disclosure, I spent a good amount of time looking for other reasons why I was not getting `connect` events to fire. In one issue or documentation somewhere, lost to the ether, I'm certain I seen something to the effect that a client connecting to a namespace also results in a connect event being triggered on `/` (since the user also joins their own room off of `/`?) 

There is a high probability my understanding is incorrect and I defer to your understanding of the protocol: I have not read myself into it.

With this said, if you want me to PR on this let me know, but I'm not clear on what the appropriate fix should be. Perhaps a `ConnectionRefusedError` originating from [_handle_connect](https://github.com/miguelgrinberg/python-socketio/blob/3bd13578c82e8a94d6b0328180606c2aefd496f1/src/socketio/server.py#L648) for `namespace` not being in `self.namespace_handlers`?

**Comment by miguelgrinberg** [comment]
> I'm certain I seen something to the effect that a client connecting to a namespace also results in a connect event being triggered on /

This used to be true, so I'm sure there are references to this behavior in old issues, but it is not done anymore.

**Comment by ba1dr** [comment]
@miguelgrinberg I understand why this has been fixed. Indeed, to prevent the server of keeping pool of potentially wrong/useless connections better to reject them in most cases. But in my case this is an internal server and all clients are trusted.
The code for message senders is in other services (perhaps could be in another repository) and web server does not need to be updated each time when I add a new namespace to the side service.
I see that I can overload method `_handle_connect` and add a handler for unknown namespace, this will fix the problem for me.

**Comment by miguelgrinberg** [comment]
Okay, I added ` namespaces` argument to the `Server` and `AsyncServer` classes. Set this to the list of namespaces you want the server to accept (in addition to any that have handlers defined). The default is `['/']`, so connections on the default namespace are accepted. Set to `'*'` (as a string, not a list) to accept all namespaces, which is the pre-5.7 behavior. Let me know if this addresses your issues.

**Comment by ba1dr** [comment]
Yes, I think this will help. Thank you for your efforts!

**Comment by miguelgrinberg** [comment]
This is now released in v5.7.1.
